ts/screens/messages/paginated/MessageRouterScreen.tsx

```diff
@@ -0,0 +1,165 @@
+  navigateToPaginatedMessageDetailScreenAction
+} from "../../../store/actions/navigation";
+import { GlobalState } from "../../../store/reducers/types";
+import { isStrictSome } from "../../../utils/pot";
+import {
+  UIMessage,
+  UIMessageDetails,
+  UIMessageId
+} from "../../../store/reducers/entities/messages/types";
+import { getDetailsByMessageId } from "../../../store/reducers/entities/messages/detailsById";
+import * as allPaginated from "../../../store/reducers/entities/messages/allPaginated";
+import {
+  Cursor,
+  getCursors
+} from "../../../store/reducers/entities/messages/allPaginated";
+
+type OwnProps = NavigationInjectedProps<{ messageId: UIMessageId }>;
+type Props = OwnProps &
+  ReturnType<typeof mapDispatchToProps> &
+  ReturnType<typeof mapStateToProps>;
+
+/**
+ * Choose the screen where to navigate, based on the message content.
+ * TODO: refactor to return an action for _dispatch_
+ * @param message
+ * @param messageDetails
+ */
+const navigateToScreenHandler =
+  (message: UIMessage, messageDetails: UIMessageDetails) =>
+  (dispatch: OwnProps["navigation"]["dispatch"]) => {
+    if (euCovidCertificateEnabled && messageDetails.euCovidCertificate) {
+      navigateBack();
+      navigateToEuCovidCertificateDetailScreen({
+        authCode: messageDetails.euCovidCertificate
+          .authCode as EUCovidCertificateAuthCode,
+        messageId: message.id
+      });
+    } else {
+      navigateBack();
+      dispatch(
+        navigateToPaginatedMessageDetailScreenAction({
+          message,
+          messageDetails
+        })
+      );
+    }
+  };
+
+const MessageRouterScreen = ({
+  cancel,
+  cursors,
+  loadMessageDetails,
+  loadPreviousPage,
+  reloadPage,
+  maybeMessage,
+  maybeMessageDetails,
+  messageId,
+  navigation
+}: Props): React.ReactElement => {
+  // used to automatically dispatch loadMessages if the pot is not some at the first rendering
+  // (avoid displaying error at the first frame)
+  const firstRendering = useRef(true);
+  const isLoading = !pot.isError(maybeMessageDetails);
+  const errorMessage = pot.isError(maybeMessageDetails)
+    ? maybeMessageDetails.error
+    : undefined;
+
+  const tryLoadMessageDetails = useCallback(() => {
+    if (maybeMessage === undefined) {
+      if (pot.isNone(cursors)) {
+        // nothing in the collection, refresh
+        reloadPage();
+      } else if (pot.isSome(cursors)) {
+        // something in the collection, get the new ones only
+        loadPreviousPage(cursors.value.previous);
+      }
+    }
+    loadMessageDetails(messageId);
+  }, [
+    maybeMessage,
+    cursors,
+    messageId,
+    loadMessageDetails,
+    loadPreviousPage,
+    reloadPage
+  ]);
+
+  useEffect(() => {
+    // message in the list and its details loaded: green light
+    if (isStrictSome(maybeMessageDetails) && maybeMessage !== undefined) {
+      navigateToScreenHandler(
+        maybeMessage,
+        maybeMessageDetails.value
+      )(navigation.dispatch);
```

Comment:
I don't know if it is the same, but actually in many other components we are using the navigation hook

```typescript
const navigation = useNavigationContext();
[...]
navigation.navigate(navigateToPayPalCheckout());
```

This is to keep the codebase homogeneous, but if it used in other parts, it's ok ;)

Reply:
In the codebase it's 26 `useNavigationContext()` vs 8 `navigation.dispatch`: using this metrics, it makes sense to follow your suggestion 😃 
On the other hand, `props.navigation` is already injected and available by the router, besides being already used by `mapStateToProps`. 

I've updated the code and its tests in this commit https://github.com/pagopa/io-app/pull/3580/commits/b9aca3b42c470f1ccacec7b626735759741c6ce8